### ISSUE TITLE
Proposal: in a class or a method, it would be nice to detect if a generic type is an enum or not

### ISSUE THREAD

**Issue by maxime-aubry** [dart-lang/language#1336]
Hello.

Today I found a limit in Dart language. I tried to detect if a generic type is an enum, whatever wich enum type it is. I did not found a way to do it.

``` dart
enum Enum1 { x, y }
enum Enum2 { x, y }

void myFunction<T>(T value) {
  if (T is enum) {
    // do something...
  } else if (T is String) {
    // do something...
  }
  // other if statements...
}

myFunction<Enum1>(Enum1.x);
myFunction<Enum2>(Enum2.y);
myFunction<String>('azerty');
```

I need that for one of my project, and I think many developers would need that too.
Thanks.
Bye.

**Comment by lrhn** [comment]
I thought I'd seen another request for a shared supertype of enum classes, but I can't find it.

Even if `enum` was a supertype of all enumerated types, `T is enum` would still be wrong (and `T is String` is also the wrong way to test whether the type `T` is a subtype of the type `String`. The `is` operator tests whether *objects* are *instances* of a type, it doesn't check relations between types. The `is` operator is also the only run-time check for type comparison available, there is no direct way to check whether one type is a subtype of another type at run-time.)

It might be better to test the *value*:
```dart
enum Enum1 { x, y }
enum Enum2 { x, y }

void myFunction<T>(T value) {
  if (value is Enum1 || value is Enum2) {
    // do something...
  } else if (value is String) {
    // do something...
  }
  // other if statements...
}

myFunction<Enum1>(Enum1.x);
myFunction<Enum2>(Enum2.y);
myFunction<String>('azerty');
```
It won't actually tell you what `T` is because `T` can be `Object` and cover all three types of values used here.

If you really want to check subtyping, you can use a helper function like:
```dart
bool isSubtype<S, T>() => <S>[] is List<T>;
```

**Comment by maxime-aubry** [comment]
Actualy, I am creating a flutter package.
Users will call methods and pass the enum type as argument.
So I can't test if it's a type or another, my package does not know them.
It must be generic, and I must use a superclass.

So, that is not possible :
``` dart
if (value is Enum1 || value is Enum2) {
    // do something...
  }
```

**Comment by lrhn** [comment]
What you are doing is not being *generic*. Dart generics (type parameters) are intended for code which works *generally* for a set of types, treating the values of all the types the same way. The code is *generic*, but the *types* are specialized with the type argument.

What you are writing here is more like *type specialization* of the *implementation*, where a function behaves *differently* depending on the type argument. That's not something the language supports directly, and you have to go through some hoops to get it. (It also tends to be somewhat fragile because it cannot know of types added in the future, and someone can validly pass `Object` as type and then you're unable to specialize the code in any way).

What I'd do, if I really had to recognize enum values, and treat values of different enum types the same way (which is itself worrisome), would be to try to read the `index` field dynamically, and catch the error you get if it fails.
```dart
  int index = -1;
  try {
     index = (value as dynamix).index;
  } on Object {
     // nope.
  }
  if (index >= 0) { // it's probably an enum.
    ...
  } else {
    // try something else.
  }
```

There currently is no relation between two enum types.

**Comment by maxime-aubry** [comment]
I will try this.
Thanks a lot!

**Comment by maxime-aubry** [comment]
Un hopefully, I does not work.
I must use the type only.
The value can be null, but I must be able to check the type.
I must instantiate an object with the type. Even if the value is null. The type must be checked.
The value can be updated later.

**Comment by maxime-aubry** [comment]
I did not find a solution yet.
Could anyone help me please ?
Remember, the value can be null. I must check if a parameter is enum only by checking the type (T).
Thanks.

**Comment by lrhn** [comment]
There is no way outside of `dart:mirrors` (and you said it's for Flutter which doesn't have mirrors) to check whether a type, either as a type variable or a `Type` object, is an enum type.
If you know all the enum types at compile-time, you can check `if (T == Enum1 || T == Enum2 ...)`, but that's the only option.


We could, as proposed before, consider introducing an `Enum` interface as a supertype of enum types. It would declare `int get index;` as the only member. It would allow you to detect that a type is *some* enum, but wouldn't help you much otherwise.

**Comment by kevmoo** [comment]
We shipped `Enum` as a base-class in 2.14 – right @lrhn ? Can close?

**Comment by lrhn** [comment]
Yes, the `Enum` class should allow `item is Enum`. We still have no way to do `T extends Enum` checks, and no plans to add them. You'll have to use the `isSubtype` function or similar, but at least you can now use `Enum` as one of the types.
